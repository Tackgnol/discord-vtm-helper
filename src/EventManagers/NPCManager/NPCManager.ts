import { find } from 'lodash';
import { settings } from '../../config/settings';
import { npcListRichEmbed, npcRichEmbed } from '../../Common';
import { Reply, ReplyType } from '../../Models/AppModels';
import { InvalidInputError } from '../../Common/Errors/InvalidInputError';
import { NPC } from '../../Models/GameData';

class NPCManager {
	displayNPCInfo(eventType: string, factSet: NPC[]): Reply {
		if (eventType === settings.subPrefixes.npcsSubCommands.all) {
			return this.allNPCs(factSet);
		} else {
			return this.oneNPC(factSet, eventType);
		}
	}

	private allNPCs(npcs: NPC[]): Reply {
		const npcList = npcs.map(npc => {
			return npc && `- ${npc.name} ${settings.lines.npcType} !vtm-npcs-${npc.callName}`;
		});
		return { type: ReplyType.Personal, value: npcListRichEmbed(npcList) };
	}

	private oneNPC(npcs: NPC[], npcCall: string): Reply {
		const npc = find(npcs, n => n.callName === npcCall);
		if (npc) {
			return { type: ReplyType.Personal, value: npcRichEmbed(npc) };
		} else {
			throw new InvalidInputError('NPC not found!');
		}
	}
}

export default NPCManager;
