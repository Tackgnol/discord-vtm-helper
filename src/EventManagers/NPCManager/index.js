const { RichEmbed } = require('discord.js');
const TurndownService = require('turndown');
const { find } = require('lodash');

class NPCManager {
	constructor(message) {
		this.message = message;
	}

	allNPCs(factSets) {
		const npcs = factSets.map(
			fact => {
				const npc = fact.npc;
				return npc && `- ${npc.name} type in !vtm-npcs-${npc.callName}`;
			})
		const richEmbed = new RichEmbed()
			.setTitle('Available NPCs')
			.setDescription(npcs);
		this.message.author.send(richEmbed);
	}

	oneNPC(factSet, npcCall) {
		let facts;
		const turndownService = new TurndownService()
		const npcs = factSet.map(
			fact => {
				const npc = fact.npc;
				return npc && { ...npc, facts: fact.facts };
			});
		const npc = find(npcs, n => n.callName === npcCall);
		if (npc) {
			const richEmbed = new RichEmbed()
				.setTitle(npc.name)
				.setImage(npc.image);

			if (npc.facts) {
				facts = npc.facts.map(
					f => (`- ${f}`)
				);
			}
			richEmbed.setDescription(turndownService.turndown(npc.description) + '\n \n' + 'Facts:\n' + facts);
			this.message.author.send(richEmbed);
		}
	}

}

module.exports = NPCManager;
