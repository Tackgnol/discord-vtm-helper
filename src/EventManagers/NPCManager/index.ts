import { find } from 'lodash';
import { settings } from '../../config/settings';
import { npcListRichEmbed, npcRichEmbed } from '../../Common/RichEmbeds';
import { Message } from 'discord.js';
import { IFactSet } from '../../Models/GameData';

class NPCManager {
	constructor(private message?: Message) {
		this.message = message;
	}

	allNPCs(factSets: IFactSet[]) {
		const npcList = factSets.map(fact => {
			const npc = fact.npc;
			return npc && `- ${npc.name} ${settings.lines.npcType} !vtm-npcs-${npc.callName}`;
		});
		const richEmbed = npcListRichEmbed(npcList);
		this.message && this.message.author.send(richEmbed);
	}

	oneNPC(factSet: IFactSet[], npcCall: string) {
		let facts;
		const npcs = factSet.map(fact => {
			const npc = fact.npc;
			return npc && { ...npc, facts: fact.facts };
		});
		const npc = find(npcs, n => n.callName === npcCall);
		if (npc) {
			const richEmbed = npcRichEmbed(npc);
			this.message && this.message.author.send(richEmbed);
		}
	}
}

export default NPCManager;
