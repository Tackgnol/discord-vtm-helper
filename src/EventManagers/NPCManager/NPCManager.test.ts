import NPCManager from './NPCManager';
import { expect } from 'chai';

describe('Managers > NPCManager test', () => {
	const npcData = [
		{
			npc: {
				name: 'test',
				description: 'this is a description',
				image: 'image',
				facts: ['fact1', 'fact2'],
				callName: 'test',
			},
			facts: ['fact1'],
		},
		{
			npc: {
				name: 'testy',
				description: 'this is a description for the second',
				image: '',
				facts: ['fact3', 'fact4'],
				callName: 'testy',
			},
			facts: ['fact4'],
		},
	];
	const npcManager = new NPCManager();
	it('returns a requested npc', () => {
		const result = npcManager.oneNPC(npcData, 'testy');
		expect(result).to.deep.eq({
			npc: {
				name: 'testy',
				description: 'this is a description for the second',
				image: '',
				facts: ['fact3', 'fact4'],
				callName: 'testy',
			},
			facts: ['fact4'],
		});
	});

	it('returns all npc on asking', () => {
		const result = npcManager.allNPCs(npcData);
		expect(result).to.deep.eq(npcData);
	});
});
