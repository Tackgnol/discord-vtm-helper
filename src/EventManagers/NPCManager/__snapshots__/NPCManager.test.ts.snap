// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventManagers >> NPCManager >> NPCManager displays a requested npc if one exists 1`] = `
Object {
  "type": 0,
  "value": Object {
    "author": null,
    "color": 5832961,
    "description": "this is a test",
    "fields": Array [
      Object {
        "inline": false,
        "name": "Facts:",
        "value": "- fact1
- fact2",
      },
    ],
    "footer": null,
    "image": null,
    "thumbnail": Object {
      "url": "placeholder",
    },
    "timestamp": null,
    "title": "Test",
    "type": "rich",
    "url": null,
  },
}
`;

exports[`EventManagers >> NPCManager >> NPCManager displays all npcs for the game 1`] = `
Object {
  "type": 0,
  "value": Object {
    "author": null,
    "color": 9044739,
    "description": "- Test call upon !vtm-npcs-testWithFacts
- Test call upon !vtm-npcs-testWithout",
    "fields": Array [],
    "footer": null,
    "image": null,
    "thumbnail": null,
    "timestamp": null,
    "title": "Speak of which thou desire knowledge",
    "type": "rich",
    "url": null,
  },
}
`;
